### PR TITLE
[Dataset] Contribute IGB-Homo dataset to node_classification.py

### PR DESCRIPTION
**Pull request by BowenYao18**
## Description
I added IGB-HOMO-dataset to the RGCN folder, currently it supports the three smaller version (tiny[~0.35GB], small[~3.5GB], medium[~35GB]).

Usage:
1. Run the "download.py --size {choose the size}", the dataset will be downloaded and processed under the default ""dataset/"" folder
2. Run the ""node_classification.py"" with the corresponding dataset you just processed

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [x] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [ ] All changes have test coverage
- [x] Code is well-documented
- [x] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [x] Related issue is referred in this PR
- [x] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

File: examples/graphbolt/download.py [hunk 1]
```diff
@@ -0,0 +1,577 @@
...
+    # array = np.memmap(file_path, dtype='float32', mode='r',  shape=(num_nodes[dataset_size][node_name], 1024))
+    array = np.load(file_path)
+    assert array.shape == (num_nodes[dataset_size][node_name], 1024)
+    gb.numpy_save_aligned(file_path, array)
+
+    # Assert the shape and elements of the array are correct
+    # new_array = np.memmap(file_path, dtype='float32', mode='r',  shape=(num_nodes[dataset_size][node_name], 1024))
+    new_array = np.load(file_path)
+    assert new_array.shape == (num_nodes[dataset_size][node_name], 1024)
+    assert np.array_equal(array, new_array)
```
Comment (mfbalin) [inline examples/graphbolt/download.py:421]: Why are these extra checks needed?

File: examples/graphbolt/node_classification.py [hunk 2]
```diff
@@ -363,7 +367,14 @@ def parse_args():
...
         type=str,
         default="ogbn-products",
-        choices=["ogbn-arxiv", "ogbn-products", "ogbn-papers100M"],
+        choices=[
+            "ogbn-arxiv",
+            "ogbn-products",
+            "ogbn-papers100M",
+            "igb-homogeneous-tiny",
+            "igb-homogeneous-small",
+            "igb-homogeneous-medium",
```
Comment (mfbalin) [inline examples/graphbolt/node_classification.py:376]: Do you think it is better to follow the convention in the IGB paper?
```suggestion
            "igb-hom-tiny",
            "igb-hom-small",
            "igb-hom-medium",
```

File: examples/graphbolt/node_classification.py [hunk 3]
```diff
@@ -309,6 +310,9 @@ def train(args, graph, features, train_set, valid_set, num_classes, model):
             # in the last layer's computation graph.
             y = data.labels
 
+            if y.dtype != "int64":
+                y = y.long()
```
Comment (mfbalin) [inline examples/graphbolt/node_classification.py:314]:
```suggestion
            y = data.labels.long()
```

Comment (mfbalin) [reply examples/graphbolt/node_classification.py:314]: This already only converts if it is not int64.

Comment (mfbalin) [reply examples/graphbolt/node_classification.py:314]: Is this conversion needed by the loss function?

File: examples/graphbolt/node_classification.py [hunk 4]
```diff
@@ -363,7 +367,14 @@ def parse_args():
...
+        choices=[
+            "ogbn-arxiv",
+            "ogbn-products",
+            "ogbn-papers100M",
+            "igb-homogeneous-tiny",
+            "igb-homogeneous-small",
+            "igb-homogeneous-medium",
+        ],
         help="The dataset we can use for node classification example. Currently"
         " ogbn-products, ogbn-arxiv, ogbn-papers100M datasets are supported.",
```
Comment (mfbalin) [inline examples/graphbolt/node_classification.py:379]:
```suggestion
        " ogbn-products, ogbn-arxiv, ogbn-papers100M and igb-hom variant datasets are supported.",
```

File: examples/graphbolt/download.py
Comment (mfbalin) [inline examples/graphbolt/download.py:1]: Can we add this as `gb.BuiltinDataset`?

Comment (BowenYao18) [reply examples/graphbolt/download.py:1]: Sure. I will ask Rui on if I can make changes in the `gb` package. It should be the `BuildnDataset` class in `on_disk_dataset.py` file?

File: examples/graphbolt/download.py (hunk 1 above)
Comment (BowenYao18) [reply examples/graphbolt/download.py:421]: These are not needed, I added them for checking if the npy files are saved correctly during testing. Also, if there is anything incorrect, it will not overwrite the original file before assert. Should I remove them?

File: examples/graphbolt/node_classification.py (hunk 3 above)
Comment (BowenYao18) [reply examples/graphbolt/node_classification.py:314]: Yes. Originally, the y have dtype of `float32` and the `crossEntropy` func will give an Exception on `expected scalar type long but found float`. Is there other ways to get around this instead of type casting? Or it is ok to do the cast?

Comment (mfbalin) [reply examples/graphbolt/node_classification.py:314]: Why not cast the labels before saving them to the file?

Comment (BowenYao18) [reply examples/graphbolt/node_classification.py:314]: Yes. I can also do that in the download.py

File: examples/graphbolt/download.py
Comment (mfbalin) [reply examples/graphbolt/download.py:1]: https://github.com/dmlc/dgl/blob/09ea3196e6c5daabf56dccecb7503aef0c3f5e69/python/dgl/graphbolt/impl/ondisk_dataset.py#L915

Comment (mfbalin) [reply examples/graphbolt/download.py:1]: You need to upload the preprocessed files to the links here: https://github.com/dmlc/dgl/blob/09ea3196e6c5daabf56dccecb7503aef0c3f5e69/python/dgl/graphbolt/impl/ondisk_dataset.py#L992-L995

Comment (mfbalin) [reply examples/graphbolt/download.py:1]: Then we don't need download.py files in your PRs at all.

File: examples/graphbolt/download.py (hunk 1 above)
Comment (mfbalin) [reply examples/graphbolt/download.py:421]: I see. The checks don't hurt us. In the end, we will upload the preprocessed folder to aws servers.

File: examples/graphbolt/download.py
Comment (mfbalin) [reply examples/graphbolt/download.py:1]: But it is best that we don't delete these files because if we notice any mistake later, we can regenerate and reprocess the dataset again.

File: examples/graphbolt/download.py [hunk 5]
```diff
@@ -0,0 +1,577 @@
...
+                "domain": "node",
+                "format": "numpy",
+                "in_memory": in_memory,
+                "name": "feat",
+                "path": "data/paper_feat.npy",
+            }
+        ],
+        "graph": {
+            "edges": [
+                {"format": "numpy", "path": "edges/paper__cites__paper.npy"}
```
Comment (mfbalin) [inline examples/graphbolt/download.py:28]: Let's add reverse edges to this file following:
https://github.com/mlcommons/training/blob/cdd928d4596c142c15a7d86b2eeadbac718c8da2/graph_neural_network/compress_graph.py#L47

File: examples/graphbolt/download.py [hunk 6]
```diff
@@ -0,0 +1,577 @@
...
+
+    data = {
+        "dataset_name": os.path.basename(path),
+        "feature_data": [
+            {
+                "domain": "node",
+                "format": "numpy",
+                "in_memory": in_memory,
+                "name": "feat",
+                "path": "data/paper_feat.npy",
```
Comment (mfbalin) [inline examples/graphbolt/download.py:23]: Is the embedding dimension for all datasets same, 1024? (homo hetero all sizes?)

Comment (BowenYao18) [reply examples/graphbolt/download.py:23]: Yes. the embedding dimension are all 1024

File: examples/graphbolt/download.py (hunk 5 above)
Comment (BowenYao18) [reply examples/graphbolt/download.py:28]: For this one, if I add rev_cite, so the there are two edgeds, it will enter the case where it is heterogeneous graph in `on_disk_dataset.py`. Is it what we want?

Comment (mfbalin) [reply examples/graphbolt/download.py]: If you check the link above, they do this:
`cites_edge = add_self_loops(remove_self_loops(paper_paper_edges)[0])[0]`
we should do the same.

Comment (mfbalin) [reply examples/graphbolt/download.py]: And then they do this: https://github.com/mlcommons/training/blob/cdd928d4596c142c15a7d86b2eeadbac718c8da2/graph_neural_network/compress_graph.py#L49

Comment (mfbalin) [reply examples/graphbolt/download.py]: But I think they made a mistake here, they should have concatenated first, then remove self loops and add self loops back.

Comment (mfbalin) [reply examples/graphbolt/download.py]: Otherwise, the self loop will be included twice.

File: examples/graphbolt/download.py [hunk 7]
```diff
@@ -0,0 +1,597 @@
...
+
+        old_edge_path = source + "/" + edge + "/" + "edge_index.npy"
+        new_edge_path = dest + "/" + edge + ".npy"
+        rev_edge_path = dest + "/" + edge + "_rev.npy"
+        os.rename(src=old_edge_path, dst=new_edge_path)
+
+        # edge_array = np.memmap(new_edge_path, dtype='int32', mode='r',  shape=(num_edges[dataset_size][edge], 2))
+        edge_array = np.load(new_edge_path)
+        new_edge_array = edge_array.transpose()
+        rev_edge_array = np.ascontiguousarray(new_edge_array[::-1])
```
Comment (mfbalin) [inline examples/graphbolt/download.py:419]: Can we remove self edges from new_edge_array, then concatenate rev_edge_array to it and then add back the self edges? We don't need 2 edge types for the homo dataset, we only need 1 edge type.

File: examples/graphbolt/download.py (hunk 5 above)
Comment (mfbalin) [reply examples/graphbolt/download.py]: To clarify, we want only 1 edge type.

File: examples/graphbolt/download.py (hunk 7 above)
Comment (BowenYao18) [reply examples/graphbolt/download.py:419]: Sure.

File: examples/graphbolt/download.py
Comment (BowenYao18) [reply examples/graphbolt/download.py:1]: Got it. I will contact Rui on how to upload the dataset

File: examples/graphbolt/download.py (hunk 7 above)
Comment (mfbalin) [reply examples/graphbolt/download.py:419]: Then we can test whether adding the reverse edges helps or hurts the accuracy for the homo case.

Comment (BowenYao18) [reply examples/graphbolt/download.py:419]: I see, I will put one paper__cites__paper file with the normal and reverse concatenated

Comment (mfbalin) [reply examples/graphbolt/download.py:419]: Let's do this (to clarify):
Remove self edges
concatenate reverse edges
add self edges

Comment (BowenYao18) [reply examples/graphbolt/download.py:419]: I tried on the tiny dataset but the 2983 classes result still is lower:
19 class: 73.00%
2983 classes: 66.22%

Comment (mfbalin) [reply examples/graphbolt/download.py:419]: Can we try with the medium? If it does not work, we can add the dataset without the reverse edges.

Comment (mfbalin) [reply examples/graphbolt/download.py:419]: And can you show me the final code?

Comment (BowenYao18) [reply examples/graphbolt/download.py:419]: Actually, the HOM dataset in the paper also have acc drop in 2K classes task comparing to the 19 class task, which looks normal and our result looks even better. However, the HET dataset they have the 2K classes tasks comparable or even better than 19 classes task, which does not make sense to me.

Comment (mfbalin) [reply examples/graphbolt/download.py:419]: Can you show results with reverse edges, without reverse edges for 19 and 2K tasks?

Comment (BowenYao18) [reply examples/graphbolt/download.py:419]: Sure. I put the code in the Slack

File: python/dgl/graphbolt/impl/ondisk_dataset.py
Comment (mfbalin) [inline python/dgl/graphbolt/impl/ondisk_dataset.py:1]: Can you add these datasets to the BuiltinDataset class docstring as well?

File: python/dgl/graphbolt/impl/ondisk_dataset.py [hunk 8]
```diff
@@ -979,6 +979,16 @@ class BuiltinDataset(OnDiskDataset):
         .. note::
             Reverse edges are added to the original graph.
 
+    **igb-hom-[tiny|small|medium]**
+        The igb-hom-[tiny|small|medium] dataset is a homogeneous citation network,
+        which is designed for developers to train and evaluate GNN models with 
+        high fidelity. See more details in `igb-hom-[tiny|small|medium] 
+        <https://mlcommons.org/2024/06/gnn-for-mlperf-training-v4/>`_.
```
Comment (mfbalin) [inline python/dgl/graphbolt/impl/ondisk_dataset.py:986]: The reference should be to the github repository of IGB datasets.

File: examples/graphbolt/node_classification.py [hunk 9]
```diff
@@ -37,6 +37,7 @@
 │
 └───> All nodes set inference & Test set evaluation
 """
+
```
Comment (mfbalin) [inline examples/graphbolt/node_classification.py:40]:
```suggestion
```

File: python/dgl/graphbolt/impl/ondisk_dataset.py (hunk 8 above)
Comment (BowenYao18) [reply python/dgl/graphbolt/impl/ondisk_dataset.py:986]: I see. i will change the URL.

**Comment by mfbalin** [comment]
What is the timeline for the large and full variants of homo?

**Comment by BowenYao18** [comment]
> What is the timeline for the large and full variants of homo?

What do you mean by "timeline"?

**Comment by mfbalin** [comment]
> > What is the timeline for the large and full variants of homo?
> 
> What do you mean by "timeline"?

In this PR, you are adding the tiny, small and medium variants. When is the next PR expected?

**Comment by BowenYao18** [comment]
> > > What is the timeline for the large and full variants of homo?
> > 
> > 
> > What do you mean by "timeline"?
> 
> In this PR, you are adding the tiny, small and medium variants. When is the next PR expected?

I don't know how long it will take for the large and full version of the hom dataset to be processed and uploaded to s3. (Since I need to reprepare the edge with self edge added) Probably need **a week** to run. If I also need to test whether the dataset is valid and runnable, it might take longer and I probably need to request a larger machine for running.

**Review by mfbalin — APPROVED** [review]
